### PR TITLE
Flesh out map. A couple iterator changes also

### PR DESCRIPTION
**Pull request by wilsonk**
Hello Alisa,

Sorry, this is a pretty large PR, but I got on a roll with map and just wanted to flesh out a good portion of it (good base functionality). There are still a few things missing when compared to C++, but, like I said, it is a good start. I may get around to adding custom comparator support, if that is something you think should be added...though I may change over the other containers to use the same API as map first?

I think merge, erase and extract may be good additions that exist in C++'s std::map also. Thoughts?

If you have any preferences, let me know, as I am enjoying this stuff. I can start on something else also, if there is some specific functionality you would like added. Please let me know if anything looks incorrect or broken (again, I only tested on windows with dmd/ldc!). I do have linux and mac machines also, so I could test that if you would like...just didn't get around to it.

Have a good day,
Kelly

### CODE REVIEW

File: src/clib/map.d [hunk 1]
```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
...
+        node = root;
+        while (node !is null) {
+            foreach (i; 0..node.filled) {
+                if (node.pairs[i].key == key) {
+                    return node.pairs[i].val;
+                }
+            }
+            node = node.next_node(key);
+        }
+        assert(0, "Internal error: newly inserted key not found");
```
Comment (nxuv) [inline src/clib/map.d]: I'd highly prefer no asserts/throws. Return null if something fails. clib in itself is more like a C library rather then C++

File: src/clib/map.d [hunk 2]
```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
...
+                }
+                index = 0;
+            } else {
+                current = null;
+                index = 0;
+            }
+        }
+
+        ref TreePair!(K, V) front() @nogc nothrow {
+            assert(!empty, "Attempting to fetch the front of an empty range");
```
Comment (nxuv) [inline src/clib/map.d]: same

File: src/clib/map.d [hunk 3]
```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
...
+            return current.pairs[index];
+        }
+
+        ref MapIterator popFront() @nogc nothrow {
+            assert(!empty, "Attempting to popFront an empty range");
```
Comment (nxuv) [inline src/clib/map.d]: same

File: src/clib/map.d [hunk 4]
```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
...
+                }
+                index = 0;
+            } else {
+                current = null;
+                index = 0;
+            }
+        }
+
+        ref const(TreePair!(K, V)) front() const @nogc nothrow {
+            assert(!empty, "Attempting to fetch the front of an empty range");
```
Comment (nxuv) [inline src/clib/map.d]: same
just a note, i think ive did something similar for error handling in list, gonna have to check

File: src/clib/map.d [hunk 5]
```diff
@@ -143,12 +366,595 @@ struct map(K, V, A: IAllocator!V = allocator!V) {
...
+            return current.pairs[index];
+        }
+
+        ref ConstMapIterator popFront() @nogc nothrow {
+            assert(!empty, "Attempting to popFront an empty range");
```
Comment (nxuv) [inline src/clib/map.d]: you get the idea

**Comment by nxuv** [comment]
You're doing damn good job. Very thankful for that
Can't work om clib rn coz of stuff but I do want it finished someday haha

Gonna check unittests on my machine today, if all good I'll merge

**Comment by wilsonk** [comment]
Sorry, I am very used to using asserts and forgot to clean them out of this code, except the unittests. I will try to remember for the future. Hopefully this change is acceptable. Tests work here on windows for dmd/ldc.

**Comment by nxuv** [comment]
lgtm, unittests pass on my machine (`just test_complete`), ty

**Pull request merged**